core-bundle/src/Resources/contao/library/Contao/StringUtil.php

```diff
@@ -582,8 +583,43 @@ static function (array $matches) use ($arrData)
 			);
 		};
 
-		$evaluateExpression = static function ($strExpression) use ($arrData)
+		// Check if we can use the expression language or if legacy tokens have been used
+		$canUseExpressionLanguage = true;
+
+		foreach (array_keys($arrData) as $token)
 		{
+			if (!preg_match('/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/', $token))
+			{
+				$canUseExpressionLanguage = false;
+				@trigger_error('Using tokens that are not valid PHP variables is deprecated. Falling back to legacy token parsing.', E_USER_DEPRECATED);
+				break;
+			}
+		}
+
+		$el = new ExpressionLanguage();
+		$el->register('constant', static function ()
+		{
+			return "throw new \\InvalidArgumentException('Cannot use the constant() function in the expression for security reasons.');";
+		}, static function ()
+		{
+			throw new \InvalidArgumentException('Cannot use the constant() function in the expression for security reasons.');
+		});
+
+		$evaluateExpression = static function ($strExpression) use ($arrData, $canUseExpressionLanguage, $el)
+		{
+			if ($canUseExpressionLanguage)
+			{
+				try
+				{
+					return $el->evaluate($strExpression, $arrData);
+				}
+				catch (\Exception $e)
+				{
+					throw new \InvalidArgumentException($e->getMessage());
```

Comment:
Why do we catch the exception and throw another one with the same message?
Couldn’t we just remove the try/catch block herer?

Reply:
The method now throws an `\InvalidArgumentException` so I wanted to convert them and do the same in case someone is catching `\InvalidArgumentException` only when calling `StringUtil::parseSimpleTokens()`.

Reply:
We should add the parent exception then I think

```suggestion
					throw new \InvalidArgumentException($e->getMessage(), 0, $e);
```

Reply:
Makes sense, done :)